### PR TITLE
ESQL: Fix filtering all elements in aggs

### PR DESCRIPTION
**Pull request by nik9000**
This adds a test to *every* agg for when it's entirely filtered away and another when filtering is enabled but unused. I'll follow up with another test later for partial filtering.

That test caught a bug where some aggs would think they'd been `seen` when they hadn't. This fixes that too.

### CODE REVIEW

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java [hunk 1]
```diff
@@ -403,6 +403,9 @@ private MethodSpec addRawVector(boolean masked) {
             if (masked) {
                 builder.beginControlFlow("if (mask.getBoolean(i) == false)").addStatement("continue").endControlFlow();
             }
+            if (stateTypeHasSeen && masked) {
+                builder.addStatement("state.seen(true)");
+            }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:408]: Another option is to build an `anyTrue` and stick it on `BooleanVector`. Then we'd check it early and bail if none are true. This'd be different than the `isConstant()` dance we do now in that it is allowed to check even. That check feels like it'd be faster than the shifted `state.seen` thing that I'm doing here....

I think I've convinced myself that's a better fix. What do you think, @ivancea?

Comment (nik9000) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:408]: I think I'll expand the microbenchmarks in the morning and check. Seems simple enough.

**Comment by nik9000** [comment]
> ```
> none   min  avgt    7  0.130 ± 0.005 -> 0.127 ±  0.001 ns/op
> ```

This is actually the most important line, though kind of obvious given the code: support for masking aggs is free if you don't use it.

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractNonThreadSafeRefCounted.java [hunk 2]
```diff
@@ -23,6 +23,7 @@ public final void incRef() {
             throw new IllegalStateException("can't increase refCount on already released object [" + this + "]");
         }
         references++;
+        // System.err.println(this + ": inc " + references);
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractNonThreadSafeRefCounted.java:26]: 👀

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java [hunk 3]
```diff
@@ -47,6 +47,9 @@ public abstract sealed class BlockHash implements Releasable, SeenGroupIds //
     /**
      * Add all values for the "group by" columns in the page to the hash and
      * pass the ordinals to the provided {@link GroupingAggregatorFunction.AddInput}.
+     * <p>
+     *     This call will not {@link GroupingAggregatorFunction.AddInput#close} {@code addInput}.
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java:51]: Were we closing them twice before? I expected another change somewhere after removing the `addInput.close()`

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java [hunk 4]
```diff
@@ -353,14 +353,14 @@ private MethodSpec addRawInput() {
...
             builder.endControlFlow();
         }
-        builder.beginControlFlow("if (mask.isConstant())");
+        builder.beginControlFlow("if (mask.allFalse())");
+        {
+            builder.addComment("Entire page masked away");
+            builder.addStatement("return");
+        }
+        builder.endControlFlow();
+        builder.beginControlFlow("if (mask.allTrue())");
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: I'm not sure this makes much sense as I understand it (?)

Before, we checked the mask in O(1), if it was all true or not. If it was all true, then we avoided checking the mask by calling `addRawVector(block)`.

Now, we check the mask in O(N), _just to avoid reading the mask later_ (Which is what we do in `addRawVector(block, mask)`). So, 2 possibilities:
1. We read every element of the mask, and it's all true. Then, we did exactly the same as in `addRawVector(block, mask)`
2. We short-circuit by finding a false. Then we're iterating the mask twice (Here and in `addRawVector`)

TL;DR: I would proably switch to `isConstant()` again (At least until we keep some constant within the blocks or something to get allTrue/False in constant time?). I don't think it's a big deal, but we're losing the improvement those ifs gave us

Comment (ivancea) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: This clearly conflicts with the benchmarks you posted. I wonder what do you think about this, or where I'm wrong here 👀

**Review by ivancea — COMMENTED** [review]
Half review here, I'll continue later! (It's mostly the tests what's missing)

File: x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/BlockHash.java:51]: Yes indeed! Sorry. I should have called this one out. We have a choice here - either class in the BlockHash or close outside. We were closing twice. Sometimes - it was inconsistent. Some BlockHashs were closing and other weren't.

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java (hunk 4 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: Crucially, `isConstant` is allowed to lie. It can return `false` if something is a constant. So it's possible that it'll all be `false` in `addRawVector(vector, mask)`. In that case we can't set `seen` to true up front - we have to set it only when we see a non-`false` value. Which costs. This checking twice avoids that.

Have a look at the commit [99bcb6a](https://github.com/elastic/elasticsearch/pull/113804/commits/99bcb6a74568cc8a90395a5d82421cad117e4c60) - that's how you have to do it if you don't have a function that can't lie.

Another option, I suppose, is for `isConstant` not to ever lie for booleans. Not hard to implement, but I'm worried that it'd cost too much at build time. And I'm not super happy with making `isConstant` do something different for booleans than for numbers. And it's quite important that numbers cal lie about being a constant. It isn't worth checking on them.

About the performance - if the block *is* a constant we get this in O(1) time still. Which is the normal case here. If you aren't using filtering you throw in a constant true block.

Comment (ivancea) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: > Crucially, isConstant is allowed to lie. It can return false if something is a constant.

I suppose by lying you mean "blocks with all the same value, that aren't a `Constant*Vector`". If that's it, indeed. Actually, the method name `isConstant` doesn't quite express that, so I guess +1 to not using it and having more semantic methods.

> In that case we can't set seen to true up front - we have to set it only when we see a non-false value.

Gotcha!

Would be nice to have a continuation or issue to improve the allTrue/False method on non-constant blocks. Not sure if worth it tho, but something to check

Comment (nik9000) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: > Actually, the method name `isConstant` doesn't quite express that, so I guess +1 to not using it and having more semantic methods.

I could rename it in the follow up to `probablyConstant`.

Comment (ivancea) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: To me, it's the "constant" part. What is a "constant" here? From what I understand, a more _semantic_ name would be "alllValuesAreTheSame()", but that's the worst method name ever. "isConstant" is a technical name, and is technically correct, as we have "Constant" blocks.
If a block was created as non-constant for _reasons_, even if all values are the same, it's not lying really. It's not a "_Constant*_" block. 
_(Anyway, philosophy, I wouldn't change anything here now)_

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java [hunk 5]
```diff
@@ -78,6 +84,15 @@ public void testBoolean() throws IOException {
                     assertThat(mask.mask().getBoolean(p), equalTo(values[p]));
                 }
             }
+            if (value != null) {
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java:87]: Using a quick Stream+toSet we would be able to check all possibilities, even if the value isn't constant. As to be as complete as possible, with not much effort (?)

Same for the other similar tests, probably

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java [hunk 6]
```diff
@@ -1168,7 +1168,9 @@ public void add(int positionOffset, IntVector groupIds) {
                 }
 
                 @Override
-                public void close() {}
+                public void close() {
+                    fail("hashes should not close AddInput");
+                }
```
Comment (ivancea) [inline x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/aggregation/blockhash/BlockHashTests.java:1173]: 🚀

**Review by ivancea — APPROVED** [review]
Second half of the review: LGTM!

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java (hunk 5 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java:87]: Checking.

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java (hunk 4 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:362]: Good names are important! I'm not going to change that method in this PR. But we can change it in another.

File: x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java (hunk 5 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java:87]: I wanted to go with something that felt less like re-implementing the feature and more like knowing up front how it was built.

Comment (ivancea) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java:87]: With that "one-liner", we get to test all `allTrue/False` methods, which are otherwise untested from what I saw.

Not a big matter, but I would reconsider doing so, maybe in another PR. As now "allTrue/False" _has to be precise_. And it would already be tested if we do some improvement in the future, like storing if there are any true/false on `add()`

Comment (nik9000) [reply x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/data/BigArrayVectorTests.java:87]: I was going to say: "this tests it 2/3 of the time which is pretty good" but then realize I have a pretty big missing step here which is when both should return `false` - which is most of that 1/3 of the time.... Ok. I'm convinced. I'll go and do it.
